### ISSUE TITLE
Command line option (for windows) to set a folder where writable files go []

### ISSUE THREAD

**Issue by wcapes** [mhx/dwarfs#176]
Not sure if possibe....

Can we have a command line option
-writable ".\settings"
where anything that is written to the file system inside of dwarfs is written to?

Thank you

**Comment by mhx** [comment]
Hi @wcapes!

I'm not sure I understand exactly what you're requesting. Which tool do you want that new option for and what is that option supposed to do?

I'm guessing that you want to mount an existing DwarFS image (using the `dwarfs` tool) and would like to be able to write to a folder within the mounted file system?

If so, what would you want to happen to files written to that folder?

**Comment by wcapes** [comment]
Hi, thanks for reply

Let me try again :)

I was thinking of packing main applications or games into the dwarfs file system ( so I can backup my personal collection, NO piracy intended ) or even have different versions of something like Total Commander.

Now Total Commander saves to an INI file and I was thinking I could pack different versions of TC into their respective dwarfs files, and have a settings folder where the settings folder would be saved.

example
![image](https://github.com/mhx/dwarfs/assets/88830951/f11b7db4-5734-4e84-9c60-738a8e52bd3c)

**Comment by wcapes** [comment]
ugh, I clicked closed by mistake, didn't mean to close it at all

**Comment by mhx** [comment]
Okay, so pretty much what I was guessing.

The question remains as to what should happen to the files written to the folder. With DwarFS being a read-only file system, the data cannot be stored to the DwarFS image. So there are a few options:

- At least on Linux, you could use something like [overlayfs](https://en.wikipedia.org/wiki/OverlayFS) to [set up a writable file system on top of a DwarFS image](https://github.com/mhx/dwarfs/blob/main/doc/dwarfs.md#setting-up-a-writable-file-system-on-top-of-a-dwarfs-image). Something like this would very likely be possible to implement on Windows using WinFSP, but I'm not aware of an existing implementation.
- Using symbolic links might work, i.e. you'd make `settings` a symbolic link pointing to a writable directory. (I've just tried this, but there seems to be at least one bug in either DwarFS or WinFSP that's preventing relative symlinks from pointing out of the mounted volume. And there's another bug, likely in DwarFS, that's preventing absolute symlinks from working.)

I'll dig into why the symbolic link approach isn't currently working. Assuming it'd work, do you think this would solve your issue?

**Comment by wcapes** [comment]
I think those files should be left there, and up to the user using the feature as to what they do with it.

To expand on my example above, one could perhaps do something like this
![image](https://github.com/mhx/dwarfs/assets/88830951/47e7f8e5-cddc-4901-b1f7-ece2c5f1467f)

That way using dwarfs Total Commander is safe, and built on the eco system, and then I can have different settings based on the version I would use

**Comment by wcapes** [comment]
The link approach could approach, need to read up on it as I've never used it before :)

**Comment by mhx** [comment]
> The link approach could approach, need to read up on it as I've never used it before :)

The symbolic link approach works fine on Linux, but WinFsp unfortunately [doesn't currently support symbolic links crossing file system boundaries](https://github.com/winfsp/winfsp/issues/454).

This isn't a limitation of DwarFS, but implementation detail of WinFsp.

I'm going to close this issue because I think there is at least one potentially viable solution that would work without any changes to DwarFS. Adding the feature to DwarFS itself would not be trivial and if I had the time I'd rather look into fixing WinFsp than adding unnecessary complexity to DwarFS.

**Comment by mhx** [comment]
Feel free to leave further comments here, though! :)

I definitely think your issue is valid, it's just that I strongly believe there's a much better solution than to add an option (and a lot of complexity) to this project.

**Comment by wcapes** [comment]
Understandable, and thank you for your time

**Comment by mhx** [comment]
FWIW, I've documented the current WinFsp issues in 79207af.

**Comment by TimeRainStarSky** [comment]
[Recent commits to WinFsp](https://github.com/LegacyAngel2K9/winfsp/commit/a67ded10544457d16a32b19f6385172969457ab4) appear to resolve this issue, though it seems DwarFS needs to add a parameter; could you please check if this solves the problem?
